lib/pagination/adapter.js

```diff
@@ -19,10 +19,21 @@ const totalPages = adapterParams => Math.ceil(totalCount(adapterParams) / adapte
 const nextPage = adapterParams =>
   adapterParams.page >= totalPages(adapterParams) ? null : adapterParams.page + 1;
 
-const pageUrl = (adapterParams, pageParam) =>
-  pageParam
-    ? `http://${adapterParams.request.headers.host}${adapterParams.request.url}?page=${pageParam}`
-    : null;
+const trimPageUrl = (originalUrl, newPageNum) => {
+  const newPageUrl = originalUrl.replace(/page=./, `page=${newPageNum}`);
+  return newPageUrl === originalUrl ? `${newPageUrl}&page=${newPageNum}` : newPageUrl;
+};
+
+const pageUrl = (adapterParams, pageParam) => {
+  if (!pageParam) return null;
+
+  const { host } = adapterParams.request.headers;
+  const { url } = adapterParams.request;
+  if (!adapterParams.request.url.includes('?')) return `http://${host}${url}?page=${pageParam}`;
+
+  const newPageUrl = trimPageUrl(url, pageParam);
+  return `http://${host}${newPageUrl}`;
```

Comment:
You assume that the protocol is always HTTP here, could be HTTPS too. Also if you have the entire request you could check in the property request.query if the param page is present or not without parsing the entire URL string (and simplify the logic related with that)

Reply:
I don't have that `query` property you mention, the object that I get here is an `IncomingMessage` and it doesn't have its query params as simple as that, I would wish it was that simple, but that's why I'm parsing everything in this "manual" way. I am aware of that issue of `http` protocol, but while I think of a solution (IncomingMessage objects doesn't have a protocol property, or the entire url, they just bring the host, the method, the url after the `/` part, and nothing more) to convert better the url, I had to put this manually for now (at least to present a valid url, otherwise would be worst).

Reply:
My mistake, I was thinking that you have an actual express request here. If you only have a raw URL you can parse it like this:
```javascript
const rawUrl = "https://my-url.com?q=1";
const parsedUrl = new URL(rawUrl);
```
Parsed URL contains something like:
```
URL {
  href: 'https://my-url.com/?q=1',
  origin: 'https://my-url.com',
  protocol: 'https:',
  username: '',
  password: '',
  host: 'my-url.com',
  hostname: 'my-url.com',
  port: '',
  pathname: '/',
  search: '?q=1',
  searchParams: URLSearchParams { 'q' => '1' },
  hash: ''
}
```

Reply:
Exactly, I don't have an actual express request because that would depend on the client that installs the paginator. About what you suggest, I also tried a workaround with the URL object, but it doesn't help me at all, because the object that I get is something like this:
```
IncomingMessage {_readableState: ReadableState, readable: true, _events: {…}, _eventsCount: 1, _maxListeners: undefined, …}
aborted: false
client: Socket {connecting: false, _hadError: false, _parent: null, _host: null, _readableState: ReadableState, …}
complete: false
connection: Socket {connecting: false, _hadError: false, _parent: null, _host: null, _readableState: ReadableState, …}
headers: {host: "127.0.0.1:50575", accept-encoding: "gzip, deflate", user-agent: "node-superagent/3.8.3", connection: "close"}
httpVersion: "1.1"
httpVersionMajor: 1
httpVersionMinor: 1
method: "GET"
rawHeaders: (8) ["Host", "127.0.0.1:50575", "Accept-Encoding", "gzip, deflate", "User-Agent", "node-superagent/3.8.3", "Connection", "close"]
rawTrailers: []
readable: true
socket: Socket {connecting: false, _hadError: false, _parent: null, _host: null, _readableState: ReadableState, …}
statusCode: null
statusMessage: null
trailers: {}
upgrade: false
url: "/index"
```
As you can see, there is nothing with the "whole url" I just have the `/` part... (what comes after the host domain), so I have to manually concat everything. URL works if I had the whole url and wanted to get parts of it, but this is the opposite case, I have to form a whole url based on the parts. If I'm missing something from the URL object that would fit this problem, let me know, but as far as I tried, URL object didn't help me at all.

Reply:
Maybe I can convert everything to URL, and then parse everything with that object, and then return it again... I didn't wanted that because at the end I would still need to convert manually the url so it's the same thing. But maybe the replace/regexp part becomes smaller... anyways I will give it another try to that URL object approach.